spec/rubocop/cop/sorbet/buggy_obsolete_strict_memoization_spec.rb

```diff
@@ -0,0 +1,149 @@
+          expect_no_offenses(<<~RUBY)
+            def foo
+              @foo = ::T.let(@foo, ::T.nilable(Foo))
+              @foo ||= Foo.new
+            end
+          RUBY
+        end
+      end
+
+      describe "with a complex type" do
+        it "doesn't register any offense" do
+          expect_no_offenses(<<~RUBY)
+            def client_info_hash
+              @client_info_hash = T.let(@client_info_hash, T.nilable(T::Hash[Symbol, T.untyped]))
+              @client_info_hash ||= client_info.to_hash
+            end
+          RUBY
+        end
+      end
+
+      describe "with multiline initialization expression" do
+        it "doesn't register any offense" do
+          expect_no_offenses(<<~RUBY)
+            def foo
+              @foo = T.let(@foo, T.nilable(Foo))
+              @foo ||= multiline_method_call(
+                foo,
+                bar,
+                baz,
+              )
+            end
+          RUBY
+        end
+
+        describe "with a gap between the two lines" do
+          it "doesn't register any offense" do
+            expect_no_offenses(<<~RUBY)
+              def foo
+                @foo = T.let(@foo, T.nilable(Foo))
+
+                @foo ||= multiline_method_call(
+                  foo,
+                  bar,
+                  baz,
+                )
+              end
+            RUBY
+          end
+        end
+      end
+
+      context "when its not the first line in a method" do
+        it "doesn't register any offense" do
+          expect_no_offenses(<<~RUBY)
+            def foo
+              some
+              other
+              code
+              @foo = T.let(@foo, T.nilable(Foo))
+              @foo ||= Foo.new
+            end
+          RUBY
+        end
+      end
+    end
+  end
+
+  describe "a mistaken variant of the obsolete memoization pattern" do
+    context "not using Sorbet" do
+      # If the project is not using Sorbet, the obsolete memoization pattern might be intentional.
+      it "does not register an offence" do
+        allow(Bundler).to(receive(:locked_gems)).and_return(
+          Struct.new(:specs).new(specs_without_sorbet),
+        )
+
+        expect_no_offenses(<<~RUBY)
+          sig { returns(Foo) }
+          def foo
+            @foo = T.let(@foo, T.nilable(Foo))
+            @foo ||= Foo.new
+          end
+        RUBY
+      end
+    end
+
+    it "registers an offence and autocorrects" do
+      # This variant would have been a mistake, which would have caused the memoized value to be discarded
+      # and recomputed on every call. We can fix it up into the working version.
+
+      expect_offense(<<~RUBY)
+        def foo
+          @foo = T.let(nil, T.nilable(Foo))
+                       ^^^ This might be a mistaken variant of the two-stage workaround that used to be needed for memoization in `#typed: strict` files. See https://sorbet.org/docs/type-assertions#put-type-assertions-behind-memoization.
```

Comment:
Kinda nice, the `^^^` is now only under the `nil` part of this, rather than the whole line.